lib/documents.js

```diff
@@ -2472,6 +2473,219 @@ function writeDocs(jobState, val, batchdef, writeBatchArray, writerId){
+  };
+
+  requestOptions.path = mlutil.databaseParam(connectionParams, path);
+  let operation = new Operation(
+      'read forestInfo', this.client, requestOptions, 'single', 'single'
+  );
+  operation.requestBody = mlutil.marshal(wrapper, operation);
+  let jobState = {
+    docInstance : this,
+    jobOptions: (jobOptions)? mlutil.copyProperties(jobOptions):{},
+    urisReadSoFar: 0,
+    urisFailedToBeRead: 0,
+    forestDict: {},
+    retryCount: 0,
+    serverTimestamp: null
+  };
+
+  if(jobState.jobOptions.batchSize){
+    if(jobState.jobOptions.batchSize > 100000) {
+      throw new Error('BatchSize cannot be greater than 100000');
+    }
+  }
+
+  jobState.jobOptions.batchSize = (jobState.jobOptions.batchSize && jobState.jobOptions.batchSize >0) ?
+      jobState.jobOptions.batchSize:1000;
+
+  jobState.stream = (jobState.jobOptions.batchSize === 1)? new stream.PassThrough():
+      new stream.PassThrough({objectMode: true});
+
+  jobState.jobOptions.queryBatchMultiple = (jobState.jobOptions.queryBatchMultiple &&
+      jobState.jobOptions.queryBatchMultiple<jobState.jobOptions.batchSize) ? jobState.jobOptions.queryBatchMultiple:null;
+
+  requester.startRequest(operation).result(mlutil.callbackOn(jobState, onQueryUrisRequest))
+      .catch((err) => operation.stream.emit('error', err));
+
+  return jobState.stream;
+}
+
+function onQueryUrisRequest(output) {
+  const jobState = this;
+  let endpoint = '/v1/internal/uris?filtered=false&start=1&pageLength='+jobState.jobOptions.batchSize;
+  jobState.query = output.query;
+  jobState.forests = output.forests;
+  jobState.requesterCount = output.forests.length;
+  jobState.startTime = Date.now();
+
+  if(jobState.jobOptions.consistentSnapshot && !jobState.serverTimestamp){
+    jobState.serverTimestamp = output.query.serverTimestamp;
+  }
+  for (let i = 0; i < jobState.forests.length; i++) {
+      jobState.forestDict[jobState.forests[i].name] = null;
+      queryAllDocs(endpoint, jobState.forests[i].name, jobState);
+  }
+}
+
+function onQueryAllDocs(output) {
+  let jobValues= this;
+  let jobState = jobValues.jobState;
+  if(output) {
+    const arr = output.split('\r\n');
+    const afterUri = arr[arr.length-2];
+    jobState.urisReadSoFar+=arr.length-1;
+    if(jobState.queryBatchMultiple){
+      const writeSize = jobState.jobOptions.batchSize/jobState.jobOptions.queryBatchMultiple;
+      let i=0;
+      while((i+writeSize)<arr.length-1){
+        const readable = stream.Readable();
+        arr.subarray(i,writeSize).forEach(item => readable.push(item));
+        readable.push(null);
+        jobState.stream.write(readable);
+        i+=writeSize;
+      }
+      if(i<arr.length-1){
+        const readable = stream.Readable();
+        arr.subarray(i,arr.length-1).forEach(item => readable.push(item));
+        readable.push(null);
+        jobState.stream.write(readable);
+      }
+    }
+    else {
+      jobState.stream.write(output);
+    }
```

Comment:
If the output is null, I believe that the callback should write the null to the stream only if the forest requester is the last active requester. 

That is, my understanding is that Node.js streams end as soon as a null is written to the stream.  So, the null shouldn't be written if other forest requesters are still active.